### DISCUSSION TITLE
A few questions about testing

### DISCUSSION THREAD

**Discussion by ckipp01** [oyvindberg/tui-scala#42] [Q&A]
So while writing some tests I noticed that I was copying quite a bit of logic that you used in here for some of the widgets. For example I used:

- Your [TestBackend.scala](https://github.com/ckipp01/skan/blob/main/skan/test/ui/TestBackend.scala) verbatim
- Your [assertBuffer method](https://github.com/ckipp01/skan/blob/main/skan/test/ui/util.scala#L103-L137)
- Your [BufferView](https://github.com/ckipp01/skan/blob/9f917998146510b9f9f5d0c82f4e24d361b8ba74/skan/test/ui/util.scala#L142-L161)

What do you think of publishing some sort of test-utils package with some of this stuff extracted out so that users could re-use some of it. Would you be open to a PR exploring this?

One thing I had a somewhat hard time with figuring out how to test was colors/modifiers. I noticed that in some of your tests you do [this sort of thing](https://github.com/oyvindberg/tui-scala/blob/93de8cb44bd3af8cb5c41b5a65d9d35f3d2f1cd3/tests/src/scala/tui/widgets/CanvasTests.scala#L20-L25) where you go over the expected ranges and apply a modifier to them. That works for small things, but having a full UI with various colors/dim/bold text gets really hard to follow that pattern with. One way around this for me was to make a copy of [diffBuffers](https://github.com/ckipp01/skan/blob/9f917998146510b9f9f5d0c82f4e24d361b8ba74/skan/test/ui/util.scala#L63-L101) with the key difference being it _only_ compares the actual symbols. Would you be open to a small change that maybe deprecates this method and adds another that has a flag that can be turned on or off to do a full diff that includes things like modifiers, vs a more minimal one that just does the symbols?

**Comment by oyvindberg** [comment]
Absolutely, it makes a lot of sense to publish some test helpers :) 

I've had the same frustrations as you're having when working with tests for this. Comparing symbols is relatively easy, and comparing modifiers and colors is quite hard. Any progress towards a better solution is good 👍 
Note that all the test patterns you're seeing are ported 1-1 from tui-rs, so I haven't really invested any time into it yet.

The thing I would really want to do is actually render the screens and compare with a checked in version, like snapshot tests. Not sure how to proceed in that direction however.
